vnext/Desktop/Modules/WebSocketModule.cpp

```diff
@@ -13,123 +14,175 @@
-                if (!optionsDynamic.empty() && !optionsDynamic["headers"].empty()) {
-                  dynamic headersDynamic = optionsDynamic["headers"];
-                  for (const auto &header : headersDynamic.items()) {
-                    options.emplace(
-                        Microsoft::Common::Unicode::Utf8ToUtf16(header.first.getString()), header.second.getString());
-                  }
-                }
-
-                this->GetOrCreateWebSocket(jsArgAsInt(args, 3), jsArgAsString(args, 0))->Connect(protocols, options);
-              }),
-          Method(
-              "close",
-              [this](dynamic args) // [int64_t code, string reason,] int64_t id
-              {
-                // See react-native\Libraries\WebSocket\WebSocket.js:_close
-                if (args.size() == 3) // WebSocketModule.close(statusCode,
-                                      // closeReason, this._socketId);
-                  this->GetOrCreateWebSocket(jsArgAsInt(args, 2))
-                      ->Close(static_cast<IWebSocketResource::CloseCode>(jsArgAsInt(args, 0)), jsArgAsString(args, 1));
-                else if (args.size() == 1) // WebSocketModule.close(this._socketId);
-                  this->GetOrCreateWebSocket(jsArgAsInt(args, 0))->Close(IWebSocketResource::CloseCode::Normal, {});
-                else {
-                  auto errorObj = dynamic::object("id", -1)("message", "Incorrect number of parameters");
-                  this->SendEvent("websocketFailed", std::move(errorObj));
-                }
-              }),
-          Method(
-              "send",
-              [this](dynamic args) // const string& message, int64_t id
-              { this->GetOrCreateWebSocket(jsArgAsInt(args, 1))->Send(jsArgAsString(args, 0)); }),
-          Method(
-              "sendBinary",
-              [this](dynamic args) // const string& base64String, int64_t id
-              { this->GetOrCreateWebSocket(jsArgAsInt(args, 1))->SendBinary(jsArgAsString(args, 0)); }),
-          Method(
-              "ping",
-              [this](dynamic args) // int64_t id
-              { this->GetOrCreateWebSocket(jsArgAsInt(args, 0))->Ping(); })};
+std::vector<facebook::xplat::module::CxxModule::Method> WebSocketModule::getMethods()
+{
+  return
+  {
+    Method(
+      "connect",
+      [this](dynamic args) // const string& url, dynamic protocols, dynamic options, int64_t id
+      {
+        IWebSocketResource::Protocols protocols;
+        dynamic protocolsDynamic = jsArgAsDynamic(args, 1);
+        if (!protocolsDynamic.empty())
+        {
+          for (const auto& protocol : protocolsDynamic.items())
+          {
+            protocols.push_back(protocol.first.getString());
+          }
+        }
+
+        IWebSocketResource::Options options;
+        dynamic optionsDynamic = jsArgAsDynamic(args, 2);
+        if (!optionsDynamic.empty() && !optionsDynamic["headers"].empty())
```

Comment:
>["headers"].empty() [](start = 54, length = 19)

operator[] inserts an empty entry if there isn't one. check that optionsDynamic.count("headers") != 0 first (meaning it's 1)